// youtube-scrapper/dist/structures/Manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Manager = void 0;
const download_1 = require("../functions/download");
const downloadFromVideo_1 = require("../functions/downloadFromVideo");
const getChannel_1 = require("../functions/getChannel");
const getPlaylistInfo_1 = require("../functions/getPlaylistInfo");
const getUser_1 = require("../functions/getUser");
const getVideoInfo_1 = require("../functions/getVideoInfo");
const getYoutubeTrending_1 = require("../functions/getYoutubeTrending");
const search_1 = require("../functions/search");
/**
 * Just a manager class with all functions this package brings to you.
 */
class Manager {
    get search() {
        return search_1.search;
    }
    get getPlaylistInfo() {
        return getPlaylistInfo_1.getPlaylistInfo;
    }
    get getVideoInfo() {
        return getVideoInfo_1.getVideoInfo;
    }
    get getUser() {
        return getUser_1.getUser;
    }
    get download() {
        return download_1.download;
    }
    get getYoutubeTrending() {
        return getYoutubeTrending_1.getYoutubeTrending;
    }
    get downloadFromVideo() {
        return downloadFromVideo_1.downloadFromVideo;
    }
    get getChannel() {
        return getChannel_1.getChannel;
    }
}
exports.Manager = Manager;
